build-tools/pause/pause.c

```diff
@@ -17,20 +17,36 @@ limitations under the License.
 #include <signal.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <sys/types.h>
+#include <sys/wait.h>
 #include <unistd.h>
 
 static void sigdown(int signo) {
-	psignal(signo, "shutting down, got signal");
-	exit(0);
+  psignal(signo, "Shutting down, got signal");
+  exit(0);
+}
+
+static void sigreap(int signo) {
+  while (waitpid(-1, NULL, WNOHANG) > 0)
+    ;
 }
 
 int main() {
-	if (signal(SIGINT, sigdown) == SIG_ERR)
-		return 1;
-	if (signal(SIGTERM, sigdown) == SIG_ERR)
-		return 2;
-	signal(SIGKILL, sigdown);
-	for (;;) pause();
-	fprintf(stderr, "error: infinite loop terminated\n");
-	return 42;
+  if (getpid() != 1)
+    fprintf(stderr, "Warning: pause should be the first process in a pod\n");
+
+  if (sigaction(SIGINT, &(struct sigaction){.sa_handler = sigdown}, NULL) < 0)
+    return 1;
+  if (sigaction(SIGTERM, &(struct sigaction){.sa_handler = sigdown}, NULL) < 0)
+    return 2;
+  if (sigaction(SIGCHLD, &(struct sigaction){.sa_handler = sigreap,
```

Comment:
There's no downside to installing a signal handler that will never get called so my preference would be to avoid special cases. If there's going to be a failure installing the handler it should happen whether or not it's run in a container.